### PR TITLE
features: terminate installs on ctrl-c and --fail-fast failures

### LINKED ISSUE #15182: Spack continues installing packages after failure
After a build failed spack will print out "Skipping build of ... since ... failed" but will continue building other packages

This occurs with KeyboardInterrupt as well

### Steps to reproduce the issue

```console
$ spack install package
(Send interrupt e.g. Ctrl-C)
...
```

### Error Message

```
 $ spack install git
==> 29647: Installing libsigsegv
==> Fetching https://ftpmirror.gnu.org/libsigsegv/libsigsegv-2.12.tar.gz
################################################################################# 100.0%
==> Staging archive: /tmp/wortiz/spack-stage/spack-stage-libsigsegv-2.12-yo4g77w4jxsh46lj4pdo3oo6szee7bjk/libsigsegv-2.12.tar.gz
==> Created stage in /tmp/wortiz/spack-stage/spack-stage-libsigsegv-2.12-yo4g77w4jxsh46lj4pdo3oo6szee7bjk
==> No patches needed for libsigsegv
==> 29647: libsigsegv: Building libsigsegv [AutotoolsPackage]
==> 29647: libsigsegv: Executing phase: 'autoreconf'
==> 29647: libsigsegv: Executing phase: 'configure'
==> Error: Exception occurred in writer daemon!
==> Error: Failed to install libsigsegv due to KeyboardInterrupt: 
Traceback (most recent call last):
  File "/home/wortiz/spack/lib/spack/llnl/util/tty/log.py", line 434, in _writer_daemon
    rlist, _, _ = select.select(istreams, [], [])
KeyboardInterrupt
==> Warning: Skipping build of m4 since libsigsegv failed
==> Warning: Skipping build of git since m4 failed
==> Warning: Skipping build of autoconf since m4 failed
==> Warning: Skipping build of automake since autoconf failed
==> Warning: Skipping build of libtool since m4 failed
==> 29647: Installing pkgconf
==> Fetching http://distfiles.dereferenced.org/pkgconf/pkgconf-1.6.3.tar.xz
################################################################################# 100.0%
==> Staging archive: /tmp/wortiz/spack-stage/spack-stage-pkgconf-1.6.3-wyezssuqshd4wismk2i4s6ppx3lilnqc/pkgconf-1.6.3.tar.xz
==> Created stage in /tmp/wortiz/spack-stage/spack-stage-pkgconf-1.6.3-wyezssuqshd4wismk2i4s6ppx3lilnqc
==> No patches needed for pkgconf
```

### Information on your system

Ubuntu 18.04
 2. any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.)

packages.yaml
```
packages:
  parmetis:
    variants: build_type=Release
  trilinos:
    variants: +teko+superlu-dist+chaco build_type=Release
  superlu-dist:
    variants: build_type=Release
  all:
    compiler: [gcc]
    providers:
      mpi: [mpich, openmpi]
      blas: [openblas, intel-mkl]
      lapack: [openblas, intel-mkl]
```

### LINKED ISSUE #15684: feature: Add option to terminate spack install after first failure
Several people have requested the ability to terminate "best effort" `spack install`'s (introduced in `v0.14.0`) after the first failure.  This feature would be useful when creating and debugging new packages.

Environment `spack install`'s continue to terminate when the installation of a package listed in the `spack.yaml` file fails.

### PR DESCRIPTION
**Pull request by tldahlgren**
Fixes #15182 
Fixes #15684 

This change restores the ability to terminate `spack install` using `ctrl-c`*.

It also adds support for a `--fail-fast` install option.  Instead of proceeding with a "best effort" installation of packages associated with a `spack install`, `spack install --fail-fast` will terminate the installation process upon detection of an install failure (i.e., by the process or another, overlapping install process).

*Note that I am able to successfully terminate an installation using `develop` by entering `ctrl-c` twice in a row.  This PR reduces the number of times it needs to be entered to one.

TODO
------
- [x] Address PR feedback
- [x] Add option to documentation

Follow-On Work
----------------
- If #15415 is merged before this PR, `environment.py` will need to be modified to support `--fast-fail`

### CODE REVIEW

**Comment by tldahlgren** [comment]
@wortiz Does this fix PR fix your `ctrl-c` problem?

**Comment by wortiz** [comment]
Yes it appears to allow `ctrl-c` to exit immediately for me

**Comment by tldahlgren** [comment]
Split the two features into two separate commits.

**Comment by tldahlgren** [comment]
@scheibelp I'm done with the split into two commits and the tests passed.

File: lib/spack/spack/installer.py [hunk 1]
```diff
@@ -1503,11 +1510,21 @@ def install(self, **kwargs):
                 raise
 
             except (Exception, KeyboardInterrupt, SystemExit) as exc:
-                # Assuming best effort installs so suppress the exception and
-                # mark as a failure UNLESS this is the explicit package.
                 err = 'Failed to install {0} due to {1}: {2}'
                 tty.error(err.format(pkg.name, exc.__class__.__name__,
                           str(exc)))
+
+                if exc.__class__.__name__ == 'KeyboardInterrupt':
```
Comment (scheibelp) [inline lib/spack/spack/installer.py]: It would be slightly clearer to create a separate `except` clause like:

```
except KeyboardInterrupt:
  raise
except Exception, SystemExit:
  ...
```

File: lib/spack/spack/test/installer.py [hunk 2]
```diff
@@ -510,6 +510,65 @@ def test_install_failed(install_mockery, monkeypatch, capsys):
...
+    with pytest.raises(KeyboardInterrupt, match=err_msg):
+        installer.install()
+
+    out = str(capsys.readouterr())
+    assert 'Failed to install' in out
+    assert err_msg in out
+
+
+def test_install_fail_fast_on_detect(install_mockery, monkeypatch, capsys):
+    """Test fail_fast install when detect an install failure."""
```
Comment (scheibelp) [inline lib/spack/spack/test/installer.py]: I think the following would read slightly clearer:

> Test fail_fast install when an install failure is detected

or

> Test that installation of packages in a DAG is suspended entirely when a single package fails to install and the fail-fast mode is enabled

Likewise, can you update the docstring of `test_install_fail_fast_on_except`?

File: lib/spack/spack/test/installer.py [hunk 3]
```diff
@@ -510,6 +510,65 @@ def test_install_failed(install_mockery, monkeypatch, capsys):
...
+    spec, installer = create_installer('a')
+
+    # Make sure the package is identified as failed
+    monkeypatch.setattr(spack.database.Database, 'prefix_failed', _true)
+
+    # Skip the actual installation though it should never get there
+    monkeypatch.setattr(inst.PackageInstaller, '_install_task', _noop)
```
Comment (scheibelp) [inline lib/spack/spack/test/installer.py]: Given that this function should not execute, I think you can either:

* remove this
* replace it with a function that throws an exception

I favor the former since we are throwing specific exception types that would indicate where the logic was terminated (and elsewhere in this review I argue that it's preferable to avoid testing an object while mocking functions inside that same object).

File: lib/spack/spack/test/installer.py [hunk 4]
```diff
@@ -510,6 +510,65 @@ def test_install_failed(install_mockery, monkeypatch, capsys):
...
+
+    msg = 'Fail-fast termination of install'
+    with pytest.raises(spack.installer.InstallError, match=msg):
+        installer.install(fail_fast=True)
+
+    out = str(capsys.readouterr())
+    assert 'b failed to install' in out
```
Comment (scheibelp) [inline lib/spack/spack/test/installer.py]: Verifying the IO output of the function makes tests more fragile (changing the messages requires changing these functions). Also, performing the tests without these checks is sufficient for verifying the desired behavior (i.e. that the installer stops iterating over the DAG when a single failure is encountered).

File: lib/spack/spack/test/installer.py [hunk 5]
```diff
@@ -510,6 +510,65 @@ def test_install_failed(install_mockery, monkeypatch, capsys):
...
+    """Test fail_fast install when encounter install failure."""
+    err_msg = 'mock install failure'
+
+    def _install(installer, task, **kwargs):
+        raise RuntimeError(err_msg)
+
+    spec, installer = create_installer('a')
+
+    # Raise a non-KeyboardInterrupt exception to trigger fast failure
+    monkeypatch.setattr(inst.PackageInstaller, '_install_task', _install)
```
Comment (scheibelp) [inline lib/spack/spack/test/installer.py]: Is it possible to have a `Package.install` implementation fail instead of mocking a `PackageInstaller` method?

Since these tests are checking the behavior of `PackageInstaller`, ideally we wouldn't have to mock `PackageInstaller` methods (this provides assurance that mocking of the methods doesn't change the behavior in ways that obfuscate the test results).

File: lib/spack/spack/installer.py [hunk 6]
```diff
@@ -1428,6 +1432,9 @@ def install(self, **kwargs):
             # assume using a separate (failed) prefix lock file.
             if pkg_id in self.failed or spack.store.db.prefix_failed(spec):
                 tty.warn('{0} failed to install'.format(pkg_id))
+                if fail_fast:
+                    raise InstallError('Fail-fast termination of install')
+
                 self._update_failed(task)
```
Comment (scheibelp) [inline lib/spack/spack/installer.py:1437]: (question) Should the `InstallError` be raised after this to alert other processes that there was a failure?

File: lib/spack/spack/installer.py (hunk 1 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: I understand.  I kept it since I wanted to have the same error message.

File: lib/spack/spack/test/installer.py (hunk 4 above)
Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: True, but it does ensure you are getting the output you expect.

File: lib/spack/spack/installer.py (hunk 1 above)
Comment (scheibelp) [reply lib/spack/spack/installer.py]: I might be misunderstanding you, but I believe my suggestion should do the same thing.

File: lib/spack/spack/test/installer.py (hunk 4 above)
Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: My opinion is that it isn't critical what *exactly* gets printed. If this corresponds somehow to the inner state of the object, I'd be inclined to argue we ought to make that measurable without checking what is printed (since the message can move around).

(For command-level tests, I think it can be important to check for output, but for object tests, I think it either provides false assurance or fragility)

File: lib/spack/spack/test/installer.py (hunk 5 above)
Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: I was narrowly focused on coverage for the `installer.py` module.  I'll have to step back and reconsider the test from the `install` level.  Code coverage is especially challenging when the functions are (still) long.

I'm not sure why mocking other methods in a class is less than ideal when the goal is to increase coverage (in challenging functions).  What am I missing?

(UPDATE: In some respects it also "encapsulates" the mocking to within the module.  I see that as a good thing.)

File: lib/spack/spack/installer.py (hunk 6 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py:1437]: Excellent point.  I debated where to place raising the error.

I rationalized that a user who wants to fail fast wants to terminate on the first package.  If there is a concurrent, overlapping installation then the worst the other install would encounter would be to try to install the package again.  If there isn't a concurrent, overlapping installation then all of the failure marks are wiped out on the next install, which has to detect the failure on its own anyway.

**Comment by tldahlgren** [comment]
@scheibelp Given your feedback involves changes to each of the separate features/commits, I'll assume that I need to split one of them -- the smaller is `ctrl-c` -- into a separate PR in order to maintain a separate commit for each.  Right?

File: lib/spack/spack/installer.py (hunk 6 above)
Comment (scheibelp) [reply lib/spack/spack/installer.py:1437]: > If there is a concurrent, overlapping installation then the worst the other install would encounter would be to try to install the package again

I agree, although I think that can also be prevented generally by adding the failure prefix. Also I'm thinking that separate processes may keep trying to install packages as long as they don't try the one that failed. Consider e.g.

```
 W
/ \
X  Y
   |
   Z
```

If there are two processes, P1, and P2, and P1 fails on X while P2 succeeds with Z, would it continue to try installing Y without the failure prefix?

**Comment by scheibelp** [comment]
> I'll assume that I need to split one of them -- the smaller is ctrl-c -- into a separate PR in order to maintain a separate commit for each. Right?

You can do that. Another option is to start by just tacking on the fixes, then (assuming the commits can be grouped cleanly) you can rebase and treat the new commits as "fixup". Both this and extracting into another PR work fine.

**Comment by tgamblin** [comment]
@tldahlgren: see https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html

**Comment by tldahlgren** [comment]
> @tldahlgren: see https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html

I've been using `git rebase -i HEAD~<#>` but guess I didn't pay enough attention as I was under the impression you could only affect the immediately preceding commit(s).

File: lib/spack/spack/installer.py (hunk 1 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: I meant the error message that provides context to the failure but I will make the change you request.

**Comment by tldahlgren** [comment]
> > @tldahlgren: see https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html
> 
> I've been using `git rebase -i HEAD~<#>` but guess I didn't pay enough attention as I was under the impression you could only affect the immediately preceding commit(s) versus "skip over" one.

Not sure what I when I started working on this again but it looks like my changes to reduce this PR to the two commits I force-pushed last week are gone, leaving a single commit for both features (and a second for a set of fixes).

File: lib/spack/spack/installer.py (hunk 6 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py:1437]: Changed

File: lib/spack/spack/test/installer.py (hunk 4 above)
Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: Removed check of `InstallError` message.

**Comment by tldahlgren** [comment]
@scheibelp ping

File: lib/spack/spack/test/installer.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: > I'm not sure why mocking other methods in a class is less than ideal when the goal is to increase coverage (in challenging functions). What am I missing?

I think increasing coverage is better. But I think a different approach should be taken. I suggested overriding `Package.install` to fail.

If you are testing a class, I think it increases complexity if you mock methods in that same class. I would draw an analogy to methods that include a `testing` parameter: if the behavior of the object is changing because of the test, how do we ensure that the object is behaving properly outside of the test environment (EDITED)? This is generally a gray area, but I think that changing the behavior of the object itself as little as possible helps reason about the test.

Does that seem reasonable?

File: lib/spack/spack/test/installer.py (hunk 4 above)
Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: This check still appears to be here. Or were you referring to removing another piece of code?

Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: I was referring to removal of one of the messages.

File: lib/spack/spack/test/installer.py (hunk 5 above)
Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: The check is in the `try-except` block within `PackageInstaller.install`'s `while` (have-build-tasks) loop.  The only things that could meaningfully be `monkeypatch`'d are the installer's `_install_task` -- or something within it (e.g., the `install` phase) -- or `_update_installed`.  

I chose to monkeypatch `_install_task` to keep the test fairly straight-forward.  As a unit test of a special case, I also wanted to keep the amount of code touched by the case as small as possible.

File: lib/spack/spack/test/installer.py (hunk 4 above)
Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: IMHO I should've been checking the message that stated the installation of `a` was being skipped since it was more meaningful so my latest (forced) push makes that change.

File: lib/spack/spack/test/installer.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: > The only things that could meaningfully be monkeypatch'd are the installer's _install_task -- or something within it (e.g., the install phase)

The install phase was what I was specifically thinking of: you can achieve that by using the mock package `failing-build` (which just throws an `InstallError` when you try to install it). You can also possibly use an entirely-mocked replacement for `Package` with an `install` method that fails.

> I chose to monkeypatch _install_task to keep the test fairly straight-forward.

I agree that it is straightforward, but I worry that it is fragile. I think it may be fragile for the reasons I explained in https://github.com/spack/spack/pull/15295/files#r391203004. 

> I also wanted to keep the amount of code touched by the case as small as possible.

I agree that is good, but I think other criteria are more important: clarity and fragility. Right now the code is clear but I'm worried about it being fragile.

To make sure: do you not see a benefit to my suggestion, or do you think the benefits of your approach outweigh the benefits of my suggestion?

Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: I don't see the fragility issue.  IMO the behavior change for the test is needed because of the structure of the code.  Also, I consider this a low-level unit test in the sense of testing a very narrow condition not easily met otherwise.

I would rather remove the test (and minimally lower coverage) than make the requested change since, IMHO, unnecessarily exercising "advanced" features like the phases in this particular case would not be worth the additional debugging effort should there be changes in that part of the code base that affect the results of this test case.

Bottom line: I am saying I think the benefits of my approach outweigh those of your suggestion.

Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: I agree that creating a test that requires knowledge of Package phases would likely be too complex. I don't think it's too complex though to make use of a package testing resource that fails. While understanding the available phases and how Spack arranges for them to execute is also complicated, I think that there will always be a `Package.install` method, and I don't think it is too much to expect that `installer.py` will always be invoking this at some point.

> I would rather remove the test (and minimally lower coverage) than make the requested change

There is some overlap between checks on `Database.prefix_failed` and logic which handles exceptions with `Package.install`, so I agree that removing this test shouldn't reduce coverage (certainly from a logic perspective if not a line-coverage perspective). I like the idea of testing both situations though.

My own summary: testing what happens when a package install fails should be simple (not just for this test but for all Spack tests). If it's not, then I should make it simple.

Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: Are you saying a basic unit test intended to ensure errors are not in the exception clause of a `try-catch` block (e.g., formats don't raise exceptions) is not as valid to retain as an integration test that you can reason about?

Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: No: I think they are both valid tests. I meant that I thought `test_install_fail_fast_on_except` and `test_install_fail_fast_on_detect` have some logical overlap and so it isn't as critical that both exist. 

As I think about it more, I think I'm wrong about that:

* `test_install_fail_fast_on_except` checks behavior when an attempt to install the package fails
* `test_install_fail_fast_on_detect` checks behavior when it is discovered that some other process failed to install the package

So in short: both tests are good and should be kept. I still think this test should change (possibly after the testing framework is refactored to make it easier to test failing package installs).

**Comment by adrienbernede** [comment]
@tldahlgren I ran into this issue today. So I am supportive of your work in this PR.

Note: on MacOS, hitting `ctrl-C` twice in a row does not stop the installation, it has the same behavior as one hit and will go to the next install.

**Comment by adrienbernede** [comment]
And I confirm that this PR works for me.

**Comment by tldahlgren** [comment]
Closing as part of re-triggering the tests.

**Comment by tldahlgren** [comment]
Re-opening to trigger the tests.

**Comment by tldahlgren** [comment]
@scheibelp @becker33 ping

**Comment by scheibelp** [comment]
Has https://github.com/spack/spack/pull/15295#discussion_r397503477 been addressed yet? I haven't seen a comment on it.

File: lib/spack/spack/test/installer.py (hunk 5 above)
Comment (becker33) [reply lib/spack/spack/test/installer.py]: @scheibelp this does not seem too fragile to me. It clearly monkeypatches a single function to fail, and that function is integral to the API, so it cannot change to break the test without complete revision of the installer, at which point rewriting tests is appropriate.

I appreciate the concern about changing behavior of an object to test it, but it is frequently the case that a "belt and suspenders" policy to programming leads to code that is very difficult to test without doing so. I do not think we should have any qualms about changing `Class.foo` as part of a test for `Class.bar`, although I agree it is appropriate for a reviewer to help ensure that the tests are properly scoped so that we do not suffer a gap in testing because of that behavior.

Given the need to push out a release soon, I think we should accept this test as-is.

Comment (tldahlgren) [reply lib/spack/spack/test/installer.py]: > The install phase was what I was specifically thinking of: you can achieve that by using the mock package `failing-build` (which just throws an `InstallError` when you try to install it). You can also possibly use an entirely-mocked replacement for `Package` with an `install` method that fails.

Using any package's overidden `install` method is not going to exercise the path that is being tested with `test_install_fail_fast_on_except`.  That path can only occur with an exception that is raised from either `_install_task` or `_update_installed`.  I will have to look at other modules I could patch.

Comment (scheibelp) [reply lib/spack/spack/test/installer.py]: It looks like the code has since been updated to resolve this.

**Comment by tldahlgren** [comment]
I will merge the last 2 commits once the tests pass.

**Comment by tldahlgren** [comment]
@becker33 @scheibelp The PR tests passed so I rebased the third commit (test change) into the second.  So, once the tests pass, this PR can be pushed to `develop` as 2 commits: one for `ctrl-c` and the other for `--fail-fast`.
